### PR TITLE
Product Gallery > Pager block: Fix block alignment

### LINKED ISSUE #10781: Product Gallery > Pager block: Fix block alignment
## Describe the bug

The Pager block must appear directly below the Large Image block and be centered with it. If we set the align property of the Product Gallery block to "wide" or "fullscreen," the Pager does not remain aligned with the Large Image block.

## To reproduce

1. From your WordPress dashboard, go to Appearance > Themes. Make sure you have a block-based theme installed and activated. If not, you can install one from the Add New option. Block-based themes include "Twenty-twenty Two," "Twenty-twenty Three," etc.
2. On the left-hand side menu, click on Appearance > Editor > Templates
3. Find and select the 'Single Product' template from the list.
4. When the Classic Product Template renders, click on Transform into Blocks. This will transform the Classic template in a block template if you haven't done it before.
5. Inside the Page editor, click on the '+' button, usually found at the top left of the editing space or within the content area itself, to add a new block.
6. In the block library that pops up, search for the 'Product Gallery' block. Click on it to add the block to the template.
7. Check that the Pager block is displayed on the Editor.
8. Select the Product Gallery block and change the Align mode on the toolbar to "Full width"
CleanShot 2023-08-30 at 14 23 39@2x

9. Check that the Pager block is not correctly centered with the Large Image block.

## Expected behavior

The Pager block must be centered and below the Large Image block.

CleanShot 2023-08-30 at 14 26 51@2x

## Screenshots

CleanShot 2023-08-30 at 14 27 37@2x

### PR DESCRIPTION
**Pull request by thealexandrelara**
## What
This PR fixes a bug that was causing the Pager block to appear misaligned with the Large Image block.

Fixes #10781

## Why
As defined in the design for the block, the Pager block must appear below the Large Image block and centered with it.

## Testing Instructions

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

1. From your WordPress dashboard, go to Appearance > Themes. Make sure you have a block-based theme installed and activated. If not, you can install one from the Add New option. Block-based themes include "Twenty-twenty Two," "Twenty-twenty Three," etc.
2. On the left-hand side menu, click on Appearance > Editor > Templates
3. Find and select the 'Single Product' template from the list.
4. When the Classic Product Template renders, click on Transform into Blocks. This will transform the Classic template in a block template if you haven't done it before.
5. Inside the Page editor, click on the '+' button, usually found at the top left of the editing space or within the content area itself, to add a new block.
6. In the block library that pops up, search for the 'Product Gallery' block. Click on it to add the block to the template.
7. Check that the Pager block is displayed on the Editor.
8. Select the Product Gallery block and change the Align mode on the toolbar to "Full width"
CleanShot 2023-08-30 at 14 23 39@2x
9. Make sure the block is correctly centered with the Large Image block.
10. Save the page and visit the product page, make sure the block is correctly being displayed and centered with the Large Image block.

* [x] Do not include in the Testing Notes 
* [ ] Should be tested by the development team exclusively 

## Screenshots or screencast

| Before | After |
| ------ | ----- |
| CleanShot 2023-08-30 at 17 04 32@2x | CleanShot 2023-08-30 at 16 42 43@2x |

## WooCommerce Visibility

Required:

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental
* [ ] N/A

### CODE REVIEW

**Review by danieldudzic — CHANGES_REQUESTED** [review]
In the editor it seems that the vertical alignment settings are broken now for the Next/Previous buttons:
HMf81J2DQg

And also, you can move the Pager up and down, not sure if this is intended:
lJx2zeGaIs

**Comment by gigitux** [comment]
It is great that E2E tests caught the regression:

image

**Comment by thealexandrelara** [comment]
Thank you for reviewing the PR and raising this points.

Initially, I had moved this block to be an inner block of the Product Gallery Large Image block; however, that caused more issues than expected so I decided to take a different approach. 

I put the block inside a group together with the Large Image block, this way we can isolate the behavior for each block. I noticed inner blocks of the Large Image block display over the image, so I think it makes sense to keep the Pager block separated from those blocks.

Finally, I had to change the way that we rearrange the blocks when a different Thumbnails position is selected and fixed the e2e tests that were failing.

**Review by danieldudzic — COMMENTED** [review]
Thanks for your work on this. This behaves as expected now.

I have one concern though.

Cause now we are moving into multiple wrapper territory:

Single_Product_‹_Template_‹_Editor_‹_newproductgallery_—_WordPress

&nbsp;<br />With the first wrapper, there was no way around it, as it needed to be set dynamically based on the `Thumbnails` position setting, but with the second one, I believe we should be able to utilize the `layout`: https://github.com/woocommerce/woocommerce-blocks/discussions/10326#discussioncomment-6544675

Did you investigate this solution? If possible I'd avoid adding another wrapper.

----------

Update: Actually scratch the above, I don't see a way we can not use a wrapper for these items:

Single_Product_‹_Template_‹_Editor_‹_newproductgallery_—_WordPress

File: assets/js/blocks/product-gallery/inner-blocks/product-gallery-pager/block.json [hunk 1]
```diff
@@ -8,6 +8,6 @@
     "category": "woocommerce",
     "keywords": [ "WooCommerce" ],
     "textdomain": "woo-gutenberg-products-block",
-	"ancestor": [ "woocommerce/product-gallery" ],
+	"ancestor": [ "woocommerce/product-gallery-large-image" ],
```
Comment (danieldudzic) [inline assets/js/blocks/product-gallery/inner-blocks/product-gallery-pager/block.json]: This doesn't seem to be correct, as now the Large Image is Pager's sibling.

Single_Product_‹_Template_‹_Editor_‹_newproductgallery_—_WordPress

**Comment by roykho** [comment]
I am seeing a console error in editor:

```
Cannot update a component (`wp`) while rendering a different component (`Edit`). To locate the bad setState() call inside `Edit`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render
```

Not related, but not sure why there is the Justification settings in two different parts which is abit confusing.

**Review by danieldudzic — APPROVED** [review]
One minor change suggestion, but overall excellent work. The updated tests also pass locally.

The **Pager** is positioned correctly, and also actually benefits from sitting inside of the container, as the user will be able to align it to the `left`, `center` and `right`:

0l9boAm1aV

**Comment by thealexandrelara** [comment]
> I am seeing a console error in editor:
> 
> ```
> Cannot update a component (`wp`) while rendering a different component (`Edit`). To locate the bad setState() call inside `Edit`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render
> ```
I cannot reproduce this, is this error appearing to you only when the Product Gallery block is inserted on the Editor?
 
> Not related, but not sure why there is the Justification settings in two different parts which is abit confusing.
> 
> 

I think this is a common behavior for the Row and Stack block :/

File: assets/js/blocks/product-gallery/inner-blocks/product-gallery-pager/block.json (hunk 1 above)
Comment (thealexandrelara) [reply assets/js/blocks/product-gallery/inner-blocks/product-gallery-pager/block.json]: Nice catch! The ancestor should be the Product Gallery block, I just updated this line of code

**Comment by roykho** [comment]
Please see:


https://github.com/woocommerce/woocommerce-blocks/assets/2132595/5e9bc553-e757-4ed5-913c-2ec5aa840c8a

**Comment by thealexandrelara** [comment]
> Please see:
> 
>  file.mov

Thank you for sharing the video, I tried multiple times to reproduce this behavior but the issue is not occurring for me :/

@danieldudzic are you experiencing this error on your side?

**Comment by danieldudzic** [comment]
> > Please see:
> > file.mov
> 
> Thank you for sharing the video, I tried multiple times to reproduce this behavior but the issue is not occurring for me :/
> 
> @danieldudzic are you experiencing this error on your side?

I also wasn't able to replicate the issue.

**Comment by gigitux** [comment]
I'm able to replicate the error reported by @roykho:

image

This error is visible in trunk as well, so I think that we can merge this PR (I didn't review it) if the blocker is the error reported. I will open an issue.

**Comment by gigitux** [comment]
Issue created https://github.com/woocommerce/woocommerce-blocks/issues/10836

**Comment by thealexandrelara** [comment]
Okay, I'll merge this PR and we can investigate and fix that issue in a separate PR. Thank you!

**Pull request merged**